hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/TokenRejectSystemContractTest.java

```diff
@@ -0,0 +1,241 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.services.bdd.suites.contract.precompile;
+
+import static com.hedera.services.bdd.junit.TestTags.SMART_CONTRACT;
+import static com.hedera.services.bdd.spec.HapiSpec.hapiTest;
+import static com.hedera.services.bdd.spec.utilops.CustomSpecAssert.allRunFor;
+import static com.hedera.services.bdd.spec.utilops.UtilVerbs.withOpContext;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.CONTRACT_REVERT_EXECUTED;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INSUFFICIENT_TOKEN_BALANCE;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_NFT_ID;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_TOKEN_ID;
+
+import com.esaulpaugh.headlong.abi.Address;
+import com.hedera.services.bdd.junit.HapiTest;
+import com.hedera.services.bdd.junit.HapiTestLifecycle;
+import com.hedera.services.bdd.junit.support.TestLifecycle;
+import com.hedera.services.bdd.spec.dsl.annotations.Account;
+import com.hedera.services.bdd.spec.dsl.annotations.Contract;
+import com.hedera.services.bdd.spec.dsl.annotations.FungibleToken;
+import com.hedera.services.bdd.spec.dsl.annotations.NonFungibleToken;
+import com.hedera.services.bdd.spec.dsl.entities.SpecAccount;
+import com.hedera.services.bdd.spec.dsl.entities.SpecContract;
+import com.hedera.services.bdd.spec.dsl.entities.SpecFungibleToken;
+import com.hedera.services.bdd.spec.dsl.entities.SpecNonFungibleToken;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.DynamicTest;
+import org.junit.jupiter.api.Tag;
+
+@Tag(SMART_CONTRACT)
+@HapiTestLifecycle
+public class TokenRejectSystemContractTest {
+
+    @Contract(contract = "TokenReject", creationGas = 1_000_000L)
+    static SpecContract tokenReject;
+
+    @Account(tinybarBalance = 1_000_000_000L)
+    static SpecAccount sender;
+
+    @BeforeAll
+    public static void setup(final @NonNull TestLifecycle lifecycle) {
+        lifecycle.doAdhoc(sender.authorizeContract(tokenReject));
+    }
+
+    @HapiTest
+    @DisplayName("Reject fungible token")
+    public Stream<DynamicTest> tokenRejectSystemContractTest(@FungibleToken SpecFungibleToken token) {
+        return hapiTest(withOpContext((spec, opLog) -> {
+            allRunFor(spec, sender.associateTokens(token), token.treasury().transferUnitsTo(sender, 100, token));
+            final var tokenAddress = token.addressOn(spec.targetNetworkOrThrow());
+            allRunFor(
+                    spec,
+                    tokenReject.call("rejectTokens", sender, new Address[] {tokenAddress}, new Address[0]),
+                    sender.getBalance().andAssert(balance -> balance.hasTokenBalance(token.name(), 0L)));
```

Comment:
would it make sense to assert if the token treasury has increase by the number of rejected tokens?

Reply:
Yes, added treasury balance checks